core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java

```diff
@@ -97,9 +97,22 @@ private int statusCode(SdkHttpFullResponse response, Optional<ExceptionMetadata>
             return response.statusCode();
         }
 
-        return modeledExceptionMetadata.filter(m -> m.httpStatusCode() != null)
-                                       .map(ExceptionMetadata::httpStatusCode)
-                                       .orElse(500);
+        if (modeledExceptionMetadata.isPresent()) {
+            ExceptionMetadata md = modeledExceptionMetadata.get();
+
+            if (md.httpStatusCode() != null) {
+                return md.httpStatusCode();
+            }
```

Comment:
Should we maybe just set the status code to 400/500 based on isFault within the code generator so that we don't need to use isFault in the runtime?

Reply:
Yeah I think that makes sense. We can push the logic all the way into `ProtocolSpec`.